private/buf/cmd/buf/command/decode/decode.go

```diff
@@ -0,0 +1,171 @@
+// Copyright 2020-2022 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package decode
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/bufbuild/buf/private/buf/bufcli"
+	"github.com/bufbuild/buf/private/bufpkg/bufanalysis"
+	"github.com/bufbuild/buf/private/bufpkg/bufreflect"
+	"github.com/bufbuild/buf/private/pkg/app/appcmd"
+	"github.com/bufbuild/buf/private/pkg/app/appflag"
+	"github.com/bufbuild/buf/private/pkg/protoencoding"
+	"github.com/bufbuild/buf/private/pkg/stringutil"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+)
+
+const (
+	errorFormatFlagName = "error-format"
+	sourceFlagName      = "source"
+	typeFlagName        = "type"
+	outputFlagName      = "output"
+	outputFlagShortName = "o"
+)
+
+// NewCommand returns a new Command.
+func NewCommand(
+	name string,
+	builder appflag.Builder,
+) *appcmd.Command {
+	flags := newFlags()
+	return &appcmd.Command{
+		Use:   name + " <descriptor>",
+		Short: "Decode binary descriptors with a source reference.",
+		Long: `The first argument is the serialized descriptor to decode.
+If no argument is specified, defaults to stdin.`,
+		Args: cobra.ExactArgs(0),
+		Run: builder.NewRunFunc(
+			func(ctx context.Context, container appflag.Container) error {
+				return run(ctx, container, flags)
+			},
+			bufcli.NewErrorInterceptor(),
+		),
+		BindFlags: flags.Bind,
+	}
+}
+
+type flags struct {
+	ErrorFormat string
+	Source      string
+	Type        string
+	Output      string
+}
+
+func newFlags() *flags {
+	return &flags{}
+}
+
+func (f *flags) Bind(flagSet *pflag.FlagSet) {
+	flagSet.StringVar(
+		&f.ErrorFormat,
+		errorFormatFlagName,
+		"text",
+		fmt.Sprintf(
+			"The format for build errors, printed to stderr. Must be one of %s.",
+			stringutil.SliceToString(bufanalysis.AllFormatStrings),
+		),
+	)
+	flagSet.StringVar(
+		&f.Source,
+		sourceFlagName,
+		"",
+		"The source that defines the serialized descriptor (e.g. buf.build/acme/weather)",
+	)
+	flagSet.StringVar(
+		&f.Type,
+		typeFlagName,
+		"",
+		"The fully-qualified type name of the serialized descriptor (e.g. acme.weather.v1.Units)",
```

Comment:
We might need to add some more information to this flag so that it's clear people can specify the fully-qualified path (including the module reference prefix).

We can work with @lucperkins on the `--help` text later on (in a separate PR) to see if we can make any other improvements.

Reply:
I added an extra line to mention the fully-qualified path, but we probably still want to make more improvements on that